### PR TITLE
✨ Gateway Registration Flow

### LINKED ISSUE #94: ✨ Gateway Registration Flow
### Checklist

- [x] I have not removed the ✨ emoji from the title.
- [x] I have searched to ensure that no existing issue covers this feature request.
- [x] For maintainers: I have updated the projects and milestones if needed.

### Description

We need to configure the gateway to handle `user.register` requests.

### Impact

This will allow users to be able to register accounts via the gateway instead of the old GameMaker monolithic server.

### Proposed Approach

1. Add a `Register` function to the `UserRpcTarget`
2. Make an API call to `UserAccount/Register`
3. Let the infrastructure layer convert the non-200 response code to the correct JSON RPC error object.

### LINKED ISSUE #172: 🧪 Tests: ApiExceptionHandler
### Checklist

- [x] I have not removed the 🧪 emoji from the title.
- [x] I have added the subject to the title (e.g. `🧪 Tests: ActorRepository`).
- [x] I have confirmed this is not already tracked in another issue.
- [x] For maintainers: I have updated the projects and milestones if needed.

### Test Type

Integration

### PR DESCRIPTION
**Pull request by mathew-odwyer**
# Description

This PR introduces the registration flow for the gateway. It ensures that the client can register an account and display any error messages where appropriate (such as API validation errors).

- Fixes #94 

## Dependencies

This PR introduces no new dependencies.

## Type of change

- [x] New feature (non-breaking change which adds functionality).

## How Has This Been Tested?

I have provided unit tests for my changes and the tests pass locally on my machine. At this stage the gateway portion of the project has almost 100% coverage (`Winterhaven.Gateway.Presentation` has 95.6% branch coverage, issues have been created to address this [here](https://github.com/mathew-odwyer/Server/issues/177) and [here](https://github.com/mathew-odwyer/Server/issues/178)).

No integration tests have been created as unit tests have covered all edge-cases. Any edge-cases outside of my changes (such as no integration tests for `UserAccountController` are a matter of API changes, not this PR).

**Test Configuration**:
* Operating System: Windows 11 Home
* Hardware: Intel i7-10710U, 32GB RAM, Intel UHD Graphics
* Toolchain: VS Community 2022

## Possible Issues

Added issues are:

- #177 
- #178 

And we also closed #172 as not planned in favour of `ApiExceptionFactory`, which has unit tests provided in this PR.

## Checklist:

- [x] I have updated the CHANGELOG to reflect my changes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have not used `ExcludeFromCodeCoverageAttribute` on a class with logic.
- [x] New and existing tests pass locally with my changes
- [x] My changes generate no new warnings

### CODE REVIEW

File: src/Gateway/Winterhaven.Gateway.Core.Application/Services/Users/IUserAccountService.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    /// <param name="emailAddress">
+    ///   The email address of the potential user to be registered.
+    /// </param>
+    /// <param name="cancellationToken">
+    ///   The cancellation token used to cancel the registration.
+    /// </param>
+    /// <returns>
+    ///   Returns a result that indicates whether the registration was successful.
+    /// </returns>
+    public Task<UserRegistrationResult> RegisterAsync(string username, string password, string emailAddress, CancellationToken cancellationToken = default);
```
Comment (mathew-odwyer) [inline src/Gateway/Winterhaven.Gateway.Core.Application/Services/Users/IUserAccountService.cs]: This shouldn't return a result, because if it doesn't throw we know we've succeeded.

File: src/Gateway/Winterhaven.Gateway.Infrastructure/Pipeline/Factories/ApiExceptionFactory.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        if (response.IsSuccessStatusCode) return null;
+
+        var problem = await TryDeserializeProblemAsync(response).ConfigureAwait(false);
+        string detail = problem?.Detail ?? "An unexpected error occurred. Please try again later.";
+
+        return response.StatusCode switch
+        {
+            HttpStatusCode.BadRequest => BuildValidationException(problem!),
+            HttpStatusCode.Unauthorized => new AuthorizationException(detail),
+            HttpStatusCode.InternalServerError => new InvalidOperationException(detail),
```
Comment (mathew-odwyer) [inline src/Gateway/Winterhaven.Gateway.Infrastructure/Pipeline/Factories/ApiExceptionFactory.cs:39]: We should throw a specific exception such as `InternalServerErrorException` to be more specific as `StreamJsonRpc` can throw `InvalidOperationException` for things such as malformed JSON-RPC requests - and being more fine-tuned here will allow us to catch the internal error in certain situations in the future for logging purposes.

File: src/Gateway/Winterhaven.Gateway.Infrastructure/Services/Users/UserAccountService.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        logger.LogInformation("Registering potential user: '{Username}'", username);
+        await userAccountClient.RegisterUserAsync(dto, cancellationToken).ConfigureAwait(false);
+        logger.LogInformation("Registered potential user: '{Username}'", username);
+
+        //// TODO: Fix this, this is a code smell
+        ////       We will always succeed unless an exception is thrown.
+        //// TODO: Figure out whether StreamJsonRpc returns anything if the return type is void in UserRpcTarget.
+        return new UserRegistrationResult(
+            Success: true);
```
Comment (mathew-odwyer) [inline src/Gateway/Winterhaven.Gateway.Infrastructure/Services/Users/UserAccountService.cs]: Fix and remove these TODOs.

**Review by mathew-odwyer — COMMENTED** [review]
Provide unit tests for:

- `UserAccountService`
- `ApiExceptionFactory`

Write issues for tests for:

- `GatewayWebSocketMessageHandler`

File: src/Gateway/Winterhaven.Gateway.Infrastructure/Pipeline/Factories/ApiExceptionFactory.cs (hunk 2 above)
Comment (mathew-odwyer) [reply src/Gateway/Winterhaven.Gateway.Infrastructure/Pipeline/Factories/ApiExceptionFactory.cs:39]: I've opted to not do this - we can instead determine where the exception was through using `when`:

`catch (InvalidOperationException ex) when (ex.Source == "StreamJsonRpc")`.
